File: src/nbody/nbody.nim
import std/[os, strutils, math]

type Body = object
  x, y, z, vx, vy, vz, mass: float64

const
  SolarMass = 4.0 * PI * PI
  DaysPerYear = 365.24

proc initBody(x, y, z, vx, vy, vz, mass: float64): Body =
  Body(x: x, y: y, z: z, vx: vx * DaysPerYear, vy: vy * DaysPerYear, vz: vz * DaysPerYear, mass: mass * SolarMass)

var bodies = [
  Body(x: 0, y: 0, z: 0, vx: 0, vy: 0, vz: 0, mass: SolarMass),
  initBody(4.84143144246472090e+00, -1.16032004402742839e+00, -1.03622044471123109e-01, 1.66007664274403694e-03, 7.69901118419740425e-03, -6.90460016972063023e-05, 9.54791938424326609e-04),
  initBody(8.34336671824457987e+00, 4.12479856412430479e+00, -4.03523417114321381e-01, -2.76742510726862411e-03, 4.99852801234917238e-03, 2.30417297573763929e-05, 2.85885980666130812e-04),
  initBody(1.28943695621391310e+01, -1.51111514016986312e+01, -2.23307578892655734e-01, 2.96460137564761618e-03, 2.37847173959480950e-03, -2.96589568540237556e-05, 4.36624404335156298e-05),
  initBody(1.53796971148509165e+01, -2.59193146099879641e+01, 1.79258772950371181e-01, 2.68067772490389322e-03, 1.62824170038242295e-03, -9.51592254519715870e-05, 5.15138902046611451e-05)
]

proc advance(bodies: var openArray[Body], dt: float64) =
  for i in 0..<bodies.len:
    for j in i+1..<bodies.len:
      let dx = bodies[i].x - bodies[j].x
      let dy = bodies[i].y - bodies[j].y
      let dz = bodies[i].z - bodies[j].z
      let d2 = dx*dx + dy*dy + dz*dz
      let mag = dt / (d2 * sqrt(d2))
      bodies[i].vx -= dx * bodies[j].mass * mag
      bodies[i].vy -= dy * bodies[j].mass * mag
      bodies[i].vz -= dz * bodies[j].mass * mag
      bodies[j].vx += dx * bodies[i].mass * mag
      bodies[j].vy += dy * bodies[i].mass * mag
      bodies[j].vz += dz * bodies[i].mass * mag
  for i in 0..<bodies.len:
    bodies[i].x += dt * bodies[i].vx
    bodies[i].y += dt * bodies[i].vy
    bodies[i].z += dt * bodies[i].vz

proc energy(bodies: openArray[Body]): float64 =
  var e = 0.0
  for i in 0..<bodies.len:
    e += 0.5 * bodies[i].mass * (bodies[i].vx*bodies[i].vx + bodies[i].vy*bodies[i].vy + bodies[i].vz*bodies[i].vz)
    for j in i+1..<bodies.len:
      let dx = bodies[i].x - bodies[j].x
      let dy = bodies[i].y - bodies[j].y
      let dz = bodies[i].z - bodies[j].z
      e -= (bodies[i].mass * bodies[j].mass) / sqrt(dx*dx + dy*dy + dz*dz)
  return e

proc offsetMomentum(bodies: var openArray[Body]) =
  var px, py, pz = 0.0
  for b in bodies:
    px += b.vx * b.mass
    py += b.vy * b.mass
    pz += b.vz * b.mass
  bodies[0].vx = -px / SolarMass
  bodies[0].vy = -py / SolarMass
  bodies[0].vz = -pz / SolarMass

let n = if paramCount() > 0: parseInt(paramStr(1)) else: 10_000_000
offsetMomentum(bodies)
echo formatFloat(energy(bodies), ffDecimal, 9)
for _ in 1..n: advance(bodies, 0.01)
echo formatFloat(energy(bodies), ffDecimal, 9)

File: src/nbody/nbody.rs
fn main() {
    let n = std::env::args().nth(1).and_then(|x| x.parse().ok()).unwrap_or(50_000_000);
    let mut bodies = [
        Body::sun(), Body::jupiter(), Body::saturn(), Body::uranus(), Body::neptune()
    ];
    let mut energy = 0.0;
    offset_momentum(&mut bodies);
    energy = calculate_energy(&bodies);
    println!("{:.9}", energy);
    for _ in 0..n {
        advance(&mut bodies, 0.01);
    }
    energy = calculate_energy(&bodies);
    println!("{:.9}", energy);
}

struct Body {
    x: f64, y: f64, z: f64,
    vx: f64, vy: f64, vz: f64,
    mass: f64,
}

const SOLAR_MASS: f64 = 4.0 * std::f64::consts::PI * std::f64::consts::PI;
const DAYS_PER_YEAR: f64 = 365.24;

impl Body {
    fn sun() -> Self {
        Body { x: 0.0, y: 0.0, z: 0.0, vx: 0.0, vy: 0.0, vz: 0.0, mass: SOLAR_MASS }
    }
    fn jupiter() -> Self {
        Body {
            x: 4.84143144246472090e+00, y: -1.16032004402742839e+00, z: -1.03622044471123109e-01,
            vx: 1.66007664274403694e-03 * DAYS_PER_YEAR, vy: 7.69901118419740425e-03 * DAYS_PER_YEAR,
            vz: -6.90460016972063023e-05 * DAYS_PER_YEAR, mass: 9.54791938424326609e-04 * SOLAR_MASS
        }
    }
    fn saturn() -> Self {
        Body {
            x: 8.34336671824457987e+00, y: 4.12479856412430479e+00, z: -4.03523417114321381e-01,
            vx: -2.76742510726862411e-03 * DAYS_PER_YEAR, vy: 4.99852801234917238e-03 * DAYS_PER_YEAR,
            vz: 2.30417297573763929e-05 * DAYS_PER_YEAR, mass: 2.85885980666130812e-04 * SOLAR_MASS
        }
    }
    fn uranus() -> Self {
        Body {
            x: 1.28943695621391310e+01, y: -1.51111514016986312e+01, z: -2.23307578892655734e-01,
            vx: 2.96460137564761618e-03 * DAYS_PER_YEAR, vy: 2.37847173959480950e-03 * DAYS_PER_YEAR,
            vz: -2.96589568540237556e-05 * DAYS_PER_YEAR, mass: 4.36624404335156298e-05 * SOLAR_MASS
        }
    }
    fn neptune() -> Self {
        Body {
            x: 1.53796971148509165e+01, y: -2.59193146099879641e+01, z: 1.79258772950371181e-01,
            vx: 2.68067772490389322e-03 * DAYS_PER_YEAR, vy: 1.62824170038242295e-03 * DAYS_PER_YEAR,
            vz: -9.51592254519715870e-05 * DAYS_PER_YEAR, mass: 5.15138902046611451e-05 * SOLAR_MASS
        }
    }
}

fn advance(bodies: &mut [Body], dt: f64) {
    for i in 0..bodies.len() {
        for j in i + 1..bodies.len() {
            let dx = bodies[i].x - bodies[j].x;
            let dy = bodies[i].y - bodies[j].y;
            let dz = bodies[i].z - bodies[j].z;
            let distance_sq = dx * dx + dy * dy + dz * dz;
            let distance = distance_sq.sqrt();
            let mag = dt / (distance_sq * distance);
            let m1 = bodies[i].mass;
            let m2 = bodies[j].mass;
            bodies[i].vx -= dx * m2 * mag;
            bodies[i].vy -= dy * m2 * mag;
            bodies[i].vz -= dz * m2 * mag;
            bodies[j].vx += dx * m1 * mag;
            bodies[j].vy += dy * m1 * mag;
            bodies[j].vz += dz * m1 * mag;
        }
    }
    for body in bodies.iter_mut() {
        body.x += dt * body.vx;
        body.y += dt * body.vy;
        body.z += dt * body.vz;
    }
}

fn calculate_energy(bodies: &[Body]) -> f64 {
    let mut energy = 0.0;
    for i in 0..bodies.len() {
        energy += 0.5 * bodies[i].mass * (bodies[i].vx * bodies[i].vx + bodies[i].vy * bodies[i].vy + bodies[i].vz * bodies[i].vz);
        for j in i + 1..bodies.len() {
            let dx = bodies[i].x - bodies[j].x;
            let dy = bodies[i].y - bodies[j].y;
            let dz = bodies[i].z - bodies[j].z;
            let distance = (dx * dx + dy * dy + dz * dz).sqrt();
            energy -= (bodies[i].mass * bodies[j].mass) / distance;
        }
    }
    energy
}

fn offset_momentum(bodies: &mut [Body]) {
    let mut px = 0.0;
    let mut py = 0.0;
    let mut pz = 0.0;
    for body in bodies.iter() {
        px += body.vx * body.mass;
        py += body.vy * body.mass;
        pz += body.vz * body.mass;
    }
    bodies[0].vx = -px / SOLAR_MASS;
    bodies[0].vy = -py / SOLAR_MASS;
    bodies[0].vz = -pz / SOLAR_MASS;
}